File: BankAccManagement.py
class Account:
    def __init__(self, account_number, holder_name, balance=0.0):
        self.__account_number = account_number
        self.__holder_name = holder_name
        self.__balance = balance

    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount
            print(f"Deposited {amount}. New balance: {self.__balance}")
        else:
            print("Invalid deposit amount.")

    def withdraw(self, amount):
        if 0 < amount <= self.__balance:
            self.__balance -= amount
            print(f"Withdrew {amount}. New balance: {self.__balance}")
        else:
            print("Invalid or insufficient amount for withdrawal.")

    def get_balance(self):
        return self.__balance

    def display_details(self):
        print(f"Account Number: {self.__account_number}")
        print(f"Holder Name: {self.__holder_name}")
        print(f"Balance: {self.__balance}")


class SavingsAccount(Account):
    def __init__(self, account_number, holder_name, balance=0.0, interest_rate=0.02):
        super().__init__(account_number, holder_name, balance)
        self.__interest_rate = interest_rate

    def apply_interest(self):
        interest = self.get_balance() * self.__interest_rate
        self.deposit(interest)
        print(f"Applied interest: {interest}. New balance: {self.get_balance()}")


class CheckingAccount(Account):
    def __init__(self, account_number, holder_name, balance=0.0, overdraft_limit=500):
        super().__init__(account_number, holder_name, balance)
        self.__overdraft_limit = overdraft_limit

    def withdraw(self, amount):
        if 0 < amount <= self.get_balance() + self.__overdraft_limit:
            new_balance = self.get_balance() - amount
            if new_balance < 0:
                print(f"Overdraft used. New balance: {new_balance}")
            super().withdraw(amount)
        else:
            print("Invalid or insufficient amount for withdrawal (including overdraft limit).")


# Example usage
savings_account = SavingsAccount(account_number="SA123", holder_name="John Doe", balance=1000)
checking_account = CheckingAccount(account_number="CA123", holder_name="Jane Doe", balance=500)

savings_account.deposit(500)
savings_account.apply_interest()
savings_account.withdraw(300)
savings_account.display_details()

checking_account.deposit(200)
checking_account.withdraw(800)
checking_account.display_details()

File: bankaccmanagement.java
import java.util.ArrayList;
import java.util.Scanner;

class Account {
    private static int counter = 1000;
    private int accountNumber;
    private String accountHolderName;
    private double balance;

    public Account(String name) {
        this.accountNumber = ++counter;
        this.accountHolderName = name;
        this.balance = 0;
    }

    public int getAccountNumber() {
        return accountNumber;
    }

    public String getAccountHolderName() {
        return accountHolderName;
    }

    public double getBalance() {
        return balance;
    }

    public void deposit(double amount) {
        if (amount > 0) {
            balance += amount;
            System.out.println("Deposited: " + amount);
        } else {
            System.out.println("Invalid deposit amount");
        }
    }

    public void withdraw(double amount) {
        if (amount > 0 && amount <= balance) {
            balance -= amount;
            System.out.println("Withdrawn: " + amount);
        } else {
            System.out.println("Insufficient balance or invalid amount");
        }
    }

    public void display() {
        System.out.println("Account Number: " + accountNumber + ", Name: " + accountHolderName + ", Balance: " + balance);
    }
}

class Bank {
    private ArrayList<Account> accounts = new ArrayList<>();

    public void createAccount(String name) {
        Account account = new Account(name);
        accounts.add(account);
        System.out.println("Account created with account number: " + account.getAccountNumber());
    }

    public Account searchAccount(int accountNumber) {
        for (Account acc : accounts) {
            if (acc.getAccountNumber() == accountNumber) {
                return acc;
            }
        }
        return null;  // Account not found
    }

    public void displayAllAccounts() {
        for (Account acc : accounts) {
            acc.display();
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Bank bank = new Bank();
        Scanner scanner = new Scanner(System.in);
        int choice;

        do {
            System.out.println("\n1. Create Account\n2. Deposit\n3. Withdraw\n4. Display Accounts\n5. Exit");
            System.out.print("Enter your choice: ");
            choice = scanner.nextInt();

            switch (choice) {
                case 1:
                    System.out.print("Enter name: ");
                    String name = scanner.next();
                    bank.createAccount(name);
                    break;

                case 2:
                    System.out.print("Enter account number: ");
                    int accNumber = scanner.nextInt();
                    Account account = bank.searchAccount(accNumber);
                    if (account != null) {
                        System.out.print("Enter amount to deposit: ");
                        double amount = scanner.nextDouble();
                        account.deposit(amount);
                    } else {
                        System.out.println("Account not found.");
                    }
                    break;

                case 3:
                    System.out.print("Enter account number: ");
                    accNumber = scanner.nextInt();
                    account = bank.searchAccount(accNumber);
                    if (account != null) {
                        System.out.print("Enter amount to withdraw: ");
                        double amount = scanner.nextDouble();
                        account.withdraw(amount);
                    } else {
                        System.out.println("Account not found.");
                    }
                    break;

                case 4:
                    bank.displayAllAccounts();
                    break;

                case 5:
                    System.out.println("Exiting...");
                    break;

                default:
                    System.out.println("Invalid choice. Please try again.");
            }
        } while (choice != 5);

        scanner.close();
    }
}
